consensus/spos/consensusMessageValidator.go

```diff
@@ -0,0 +1,270 @@
+			len(cnsMsg.Signature))
+	}
+
+	isNodeInEligibleList := wrk.consensusState.IsNodeInEligibleList(string(cnsMsg.PubKey))
+	if !isNodeInEligibleList {
+		return fmt.Errorf("%w : received message from consensus topic has an invalid public key: %s",
+			ErrNodeIsNotInEligibleList,
+			logger.DisplayByteSlice(cnsMsg.PubKey))
+	}
+
+	if wrk.consensusState.RoundIndex+1 < cnsMsg.RoundIndex {
+		log.Trace("received message from consensus topic has a future round",
+			"msg type", wrk.consensusService.GetStringValue(msgType),
+			"from", cnsMsg.PubKey,
+			"header hash", cnsMsg.BlockHeaderHash,
+			"msg round", cnsMsg.RoundIndex,
+			"round", wrk.consensusState.RoundIndex,
+		)
+
+		return ErrMessageForFutureRound
+	}
+
+	if wrk.consensusState.RoundIndex > cnsMsg.RoundIndex {
+		log.Trace("received message from consensus topic has a past round",
+			"msg type", wrk.consensusService.GetStringValue(msgType),
+			"from", cnsMsg.PubKey,
+			"header hash", cnsMsg.BlockHeaderHash,
+			"msg round", cnsMsg.RoundIndex,
+			"round", wrk.consensusState.RoundIndex,
+		)
+
+		return ErrMessageForPastRound
+	}
+
+	err = wrk.checkSignature(cnsMsg)
+	if err != nil {
+		return fmt.Errorf("%w : verify signature for received message from consensus topic failed: %s",
+			ErrInvalidSignature,
+			err.Error())
+	}
+
+	return nil
+}
+
+func (wrk *Worker) checkConsensusMessageValidityForMessageType(cnsMsg *consensus.Message) error {
+	msgType := consensus.MessageType(cnsMsg.MsgType)
+
+	if wrk.consensusService.IsMessageWithBlockBodyAndHeader(msgType) {
+		return wrk.checkMessageWithBlockBodyAndHeaderValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithBlockBody(msgType) {
+		return wrk.checkMessageWithBlockBodyValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithBlockHeader(msgType) {
+		return wrk.checkMessageWithBlockHeaderValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithSignature(msgType) {
+		return wrk.checkMessageWithSignatureValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithFinalInfo(msgType) {
+		return wrk.checkMessageWithFinalInfoValidity(cnsMsg)
+	}
+
+	return fmt.Errorf("%w : received message type from consensus topic is invalid: %d",
+		ErrInvalidMessageType,
+		msgType)
+}
+
+func (wrk *Worker) checkMessageWithBlockBodyAndHeaderValidity(cnsMsg *consensus.Message) error {
+	isMessageInvalid := cnsMsg.SignatureShare != nil ||
+		cnsMsg.PubKeysBitmap != nil ||
+		cnsMsg.AggregateSignature != nil ||
+		cnsMsg.LeaderSignature != nil
+
+	if isMessageInvalid {
+		return fmt.Errorf("%w : received message from consensus topic is invalid: "+
+			"SignatureShare = %s PubKeysBitmap = %s AggregateSignature = %s LeaderSignature = %s",
+			ErrInvalidMessage,
+			logger.DisplayByteSlice(cnsMsg.SignatureShare),
+			logger.DisplayByteSlice(cnsMsg.PubKeysBitmap),
+			logger.DisplayByteSlice(cnsMsg.AggregateSignature),
+			logger.DisplayByteSlice(cnsMsg.LeaderSignature))
+	}
+
+	if len(cnsMsg.Body) > core.MegabyteSize {
```

Comment:
This could not happen

Reply:
It couldn't happen in the current setting of the p2p, but if in the future this limit would be changed, this condition would prevent spam on this field greater than 1MB

Reply:
If we allow p2p messages of 2 or more MB, we certainly want a larger header to be processed.

Reply:
Depends on the time which will take the broadcast header to reach to every node in a round time frame